src/mbgl/mtl/texture2d.cpp

```diff
@@ -75,126 +70,103 @@ size_t Texture2D::numChannels() const noexcept {
     }
 }
 
-/*
+MTL::PixelFormat Texture2D::getMetalPixelFormat() const noexcept {
+    switch (channelType) {
+        case gfx::TextureChannelDataType::UnsignedByte:
+            switch (pixelFormat) {
+                case gfx::TexturePixelType::RGBA:
+                    return MTL::PixelFormat::PixelFormatRGBA8Unorm;
+                case gfx::TexturePixelType::Alpha:
+                    return MTL::PixelFormat::PixelFormatA8Unorm;
+                default:
+                    assert(false);
+            }
+        case gfx::TextureChannelDataType::HalfFloat:
+            switch (pixelFormat) {
+                case gfx::TexturePixelType::RGBA:
+                    return MTL::PixelFormat::PixelFormatRGBA16Float;
+                case gfx::TexturePixelType::Alpha:
+                    return MTL::PixelFormat::PixelFormatR16Float;
+                default:
+                    assert(false);
+            }
+    }
+}
+
 void Texture2D::createObject() noexcept {
     // Create a new texture object
-    assert(!textureResource);
-    auto obj = context.createUniqueTexture();
-    const auto storageSize = gl::TextureResource::getStorageSize(size, pixelFormat, channelType);
-    context.renderingStats().memTextures += storageSize;
+    assert(!metalTexture.get());
+    auto textureDescriptor = NS::RetainPtr(
+        MTL::TextureDescriptor::texture2DDescriptor(getMetalPixelFormat(), size.width, size.height, false));
+    metalTexture = context.createMetalTexture(textureDescriptor);
 
-    // @TODO: TextureResource is still needed while we have legacy rendering pathways
-    textureResource = std::make_unique<gl::TextureResource>(std::move(obj), storageSize);
+    context.renderingStats().memTextures += getDataSize();
 }
 
 void Texture2D::createStorage(const void* data) noexcept {
-    assert(textureResource);
-
-    // Create backing storage for our texture object
-    using namespace platform;
-
-    // Bind to TU 0 and upload
-    context.activeTextureUnit = 0;
-    context.texture[0] = getTextureID();
-    context.pixelStoreUnpack = {1};
-    MBGL_CHECK_ERROR(glTexImage2D(GL_TEXTURE_2D,
-                                  0,
-                                  Enum<gfx::TexturePixelType>::sizedFor(pixelFormat, channelType),
-                                  size.width,
-                                  size.height,
-                                  0,
-                                  Enum<gfx::TexturePixelType>::to(pixelFormat),
-                                  Enum<gfx::TextureChannelDataType>::to(channelType),
-                                  data));
+    assert(metalTexture.get());
+
+    MTL::Region region = MTL::Region::Make2D(0, 0, size.width, size.height);
+    NS::UInteger bytesPerRow = size.width * getPixelStride();
+    metalTexture->replaceRegion(region, 0, data, bytesPerRow);
```

Comment:
Since this is pretty much identical to `uploadSubRegion`, I think we can drop `createStorage` here.